refactor(core): Handle corrupted text nodes during serialization and hydration process

jessicajaniuk opened the pull request:
This PR handles two cases:
- Restoring empty text nodes that were lost during serialization
- Preserving text nodes in the case that a text node is immediately adjacent to another text node

Both are cases in which information about those nodes / separators being present is lost when serialized and transferred to the browser. We handle this with special comment nodes that are added just prior to SSR DOM serialization. Those comment nodes are found and replaced with their respective empty text nodes or separations during the hydration process.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
refactor(core): Handle cases of lost empty text nodes during hydration

During DOM serialization, empty text nodes are lost when the client parses the DOM. To solve this problem comment nodes are added where the empty nodes are located right before serialization. Those comments then get replaced during hydration with the proper empty text nodes.

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -135,6 +139,41 @@ function serializeLView(lView: LView, context: HydrationContext): SerializedView
+      } else {
+        // Handle cases where text nodes can be lost after DOM serialization:
+        //  1. When there is an *empty text node* in DOM: in this case, this
+        //     node would not make it into the serialized string and as s result,
```
```suggestion
        //     node would not make it into the serialized string and as a result,
```

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -83,8 +85,10 @@ export function annotateForHydration(appRef: ApplicationRef, doc: Document) {
+          corruptedTextNodes,
         };
         annotateHostElementForHydration(hostElement as HTMLElement, lView, context);
+        insertTextNodeMarkers(corruptedTextNodes, doc);
```
```suggestion
        insertCorruptedTextNodeMarkers(corruptedTextNodes, doc);
```

AndrewKushnir commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -52,6 +52,7 @@ class SerializedViewCollection {
  */
 interface HydrationContext {
   serializedViewCollection: SerializedViewCollection;
+  corruptedTextNodes: Map<HTMLElement, string>;
```
I think it'd be great to add a JSDoc comment above this to explain what the map is used for (a couple of sentences).

AndrewKushnir replied:
We can also use an enum to better represent the operation and its type:
```suggestion
  corruptedTextNodes: Map<HTMLElement, TextNodeMarker>;
```
where `TextNodeMarker` may look like this:
```
export const enum TextNodeMarker {
  EmptyNode = 'ngetn';
  Separator = 'ngtns';
}
```

jessicajaniuk replied:
This is great idea. I didn't even think about an enum.

AndrewKushnir commented on packages/core/src/hydration/utils.ts:
```diff
@@ -123,6 +146,72 @@ export function getComponentLViewForHydration(viewRef: ViewRef): LView|null {
+ */
+export function processTextNodeMarkersBeforeHydration(node: HTMLElement) {
+  const doc = getDocument();
+  const commentIterator = doc.createNodeIterator(node, NodeFilter.SHOW_COMMENT, {
```
```suggestion
  const commentNodesIterator = doc.createNodeIterator(node, NodeFilter.SHOW_COMMENT, {
```

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -662,26 +662,36 @@ function createViewBlueprint(bindingStartIndex: number, initialViewLength: numbe
+
+      // When using native Shadow DOM, do not clear host element to allow native slot
+      // projection.
+      const preserveContent = preserveHostContent || encapsulation === ViewEncapsulation.ShadowDom;
+      return renderer.selectRootElement(elementOrSelector, preserveContent);
+    };
```
We can avoid patching the whole `locateRootElement` function and instead update an existing one to add one more call there:

```
--- a/packages/core/src/render3/instructions/shared.ts
+++ b/packages/core/src/render3/instructions/shared.ts
@@ -682,7 +682,9 @@ export function locateHostElement(
   // When using native Shadow DOM, do not clear host element to allow native slot
   // projection.
   const preserveContent = preserveHostContent || encapsulation === ViewEncapsulation.ShadowDom;
-  return renderer.selectRootElement(elementOrSelector, preserveContent);
+  const rootElement = renderer.selectRootElement(elementOrSelector, preserveContent);
+  applyRootElementTransform(rootElement);
+  return rootElement;
 }
```

and update the `applyRootElementTransform` to change from the `(rootElement: RElement) => {}` (the default one, noop when hydration is not enabled) to `(rootElement: RElement) => processTextNodeMarkersBeforeHydration(RElement)` via the switch approach that we use in other places.

jessicajaniuk replied:
I've updated this to use the suggested implementation.

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+        verifyClientAndSSRContentsMatch(ssrContents, clientRootNode);
+      });
+
+      xit('should support projected text node content with plain text nodes', async () => {
```
Let's add a "FIXME(akushnir): enable it once hydration for content projection is implemented" above it, so that we enable it later.

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+        const html = await ssr(SimpleComponent);
+        const ssrContents = getAppContents(html);
+
+        // TODO: properly assert `ngh` attribute value once the `ngh`
+        // format stabilizes, for now we just check that it's present.
```
We can drop this TODO for now (we'll see if we want to have better asserts for generated annotations later).

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+          `,
+        })
+        class SimpleComponent {
+          spanText = ''
+          pText = ''
+          anotherText = ''
```
```suggestion
          spanText = '';
          pText = '';
          anotherText = '';
```

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+          standalone: true,
+          selector: 'app',
+          template: `
+            This is hydrated content.<span>{{spanText}}</span>.
```
```suggestion
            This is hydrated content.
            <span>{{spanText}}</span>.
```

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
          });
     });
+
+    describe('text nodes', () => {
```
```suggestion
    describe('corrupted text nodes restoration', () => {
```

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+        verifyClientAndSSRContentsMatch(ssrContents, clientRootNode);
+      });
+
+      it('should support empty text nodes with subsequent content', async () => {
```
```suggestion
      it('should support restoration of multiple text nodes in a row', async () => {
```

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+          `,
+        })
+        class SimpleComponent {
+          emptyText = ''
+          moreText = ''
+          andMoreText = ''
+          secondEmptyText = ''
+          secondMoreText = ''
```
```suggestion
          emptyText = '';
          moreText = '';
          andMoreText = '';
          secondEmptyText = '';
          secondMoreText = '';
```

AndrewKushnir commented on packages/core/src/hydration/utils.ts:
```diff
@@ -123,6 +146,72 @@ export function getComponentLViewForHydration(viewRef: ViewRef): LView|null {
+  const doc = getDocument();
+  const commentIterator = doc.createNodeIterator(node, NodeFilter.SHOW_COMMENT, {
+    acceptNode(node) {
+      const content = node.textContent;
```
I'd propose doing the trimming here as well (similar to what we do during serialization, may be extracting this into a separate fn, like `getTextNodeContent`):
```suggestion
      const content = node.textContent.replace(/\s/gm, '');
```

AndrewKushnir commented on packages/core/src/hydration/utils.ts:
```diff
@@ -123,6 +146,72 @@ export function getComponentLViewForHydration(viewRef: ViewRef): LView|null {
+    }
+  });
+  let currentNode: Comment;
+  // We cannot modify the DOM while using the commentIterator
+  // So we have to grab each node and then follow up with a
+  // replacement of the node with the new empty text node.
```
```suggestion
  // We cannot modify the DOM while using the commentIterator,
  // because it throws off the iterator state.
  // So we collect all marker nodes first and then follow up with
  // applying the changes to the DOM: either inserting an empty node
  // or just removing the marker if it was used as a separator.
```

AndrewKushnir commented on packages/platform-server/test/hydration_spec.ts:
```diff
@@ -903,5 +911,105 @@ describe('platform-server integration', () => {
+          selector: 'app',
+          imports: [NgIf],
+          template: `
+            <div>Hello <ng-container *ngIf="true">Angular</ng-container> <ng-container *ngIf="true">World</ng-container></div>
```
(formatting)
```suggestion
            <div>
              Hello
              <ng-container *ngIf="true">Angular</ng-container>
              <ng-container *ngIf="true">World</ng-container>
            </div>
```

thePunderWoman committed:
refactor(core): Handle lost adjacent text nodes during hydration

This restores the separation between adjacent text nodes that is lost during server serialization when parsed by the browser. It adds a special comment node just prior to the serialization process that is then restored as a separated node during hydration.

AndrewKushnir reviewed (approved):
@jessicajaniuk looks great 👍

alxhub commented on packages/core/src/hydration/utils.ts:
```diff
@@ -123,6 +145,48 @@ export function getComponentLViewForHydration(viewRef: ViewRef): LView|null {
+  while (currentNode = commenNodestIterator.nextNode() as Comment) {
+    nodes.push(currentNode);
+  }
+  for (let node of nodes) {
```
```suggestion
  for (const node of nodes) {
```

alxhub commented on packages/core/src/hydration/utils.ts:
```diff
@@ -123,6 +145,48 @@ export function getComponentLViewForHydration(viewRef: ViewRef): LView|null {
+ */
+export function processTextNodeMarkersBeforeHydration(node: HTMLElement) {
+  const doc = getDocument();
+  const commenNodestIterator = doc.createNodeIterator(node, NodeFilter.SHOW_COMMENT, {
```
```suggestion
  const commentNodesIterator = doc.createNodeIterator(node, NodeFilter.SHOW_COMMENT, {
```

alxhub commented on packages/core/src/hydration/annotate.ts:
```diff
@@ -243,3 +282,19 @@ function annotateHostElementForHydration(
+ */
+function insertCorruptedTextNodeMarkers(
+    corruptedTextNodes: Map<HTMLElement, string>, doc: Document) {
+  for (let [textNode, marker] of corruptedTextNodes) {
```
```suggestion
  for (const [textNode, marker] of corruptedTextNodes) {
```

jessicajaniuk commented:
TGP is green

alxhub commented:
This PR was merged into the repository by commit 81e974a99d0a4d30cfcfc53ad3e15484925c8fa0.
